File: code/python/05-context-assembly/condition_engine.py
"""Condition Engine — evaluate DSL conditions for dynamic prompt sections.

Supports a simple, safe condition DSL for including or excluding prompt
sections based on runtime variables.  No ``eval()`` is used; conditions are
parsed with regex-based tokenisation.

Simple conditions::

    "plan == 'premium'"
    "user.plan in ['premium', 'enterprise']"
    "sentiment_score > 0.7"
    "user.email contains '@enterprise'"
    "conversation_history exists"

Compound conditions::

    "plan == 'premium' AND country == 'US'"
    "country == 'DE' OR country == 'FR'"

See: docs/04-context-engineering/02-dynamic-prompt-assembly.md
"""

from __future__ import annotations

import ast
import logging
import re
from typing import Any

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Operator table
# ---------------------------------------------------------------------------

_OP_TABLE: dict[str, Any] = {
    "eq":       lambda a, b: a == b,
    "neq":      lambda a, b: a != b,
    "in":       lambda a, b: a in b,
    "not_in":   lambda a, b: a not in b,
    "gt":       lambda a, b: a > b,
    "lt":       lambda a, b: a < b,
    "gte":      lambda a, b: a >= b,
    "lte":      lambda a, b: a <= b,
    "contains": lambda a, b: b in a,
    "exists":   lambda a, _: a is not None,
}

# Operator regex patterns — order matters: longer/multi-word first
_OP_PATTERNS: list[tuple[str, str]] = [
    (r"not\s+in",     "not_in"),
    (r"not_in",       "not_in"),
    (r">=",           "gte"),
    (r"<=",           "lte"),
    (r"==",           "eq"),
    (r"!=",           "neq"),
    (r">",            "gt"),
    (r"<",            "lt"),
    (r"\bcontains\b", "contains"),
    (r"\bexists\b",   "exists"),
    (r"\bin\b",       "in"),
]

_AND_SPLITTER = re.compile(r"\bAND\b")
_OR_SPLITTER  = re.compile(r"\bOR\b")


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _get_nested(variables: dict, key: str) -> Any:
    """Resolve a dotted key path: ``'user.plan'`` → ``variables['user']['plan']``."""
    obj: Any = variables
    for part in key.split("."):
        if isinstance(obj, dict):
            obj = obj.get(part)
        else:
            obj = getattr(obj, part, None)
        if obj is None:
            return None
    return obj


def _parse_rhs(raw: str | None) -> Any:
    """Parse the RHS of a condition into a Python literal (str, int, float, list …)."""
    if raw is None:
        return None
    raw = raw.strip()
    try:
        return ast.literal_eval(raw)
    except (ValueError, SyntaxError):
        return raw


def _parse_atom(atom_str: str) -> tuple[str, str, Any]:
    """Parse one atomic condition: ``'key op value'`` → ``(key, op_name, value)``.

    Args:
        atom_str: A single condition expression, e.g. ``"user.plan == 'premium'"``.

    Returns:
        Tuple of *(key, op_name, parsed_value)*.

    Raises:
        ValueError: If the condition cannot be parsed.
    """
    atom_str = atom_str.strip()
    for op_re, op_name in _OP_PATTERNS:
        pattern = rf"^([\w.]+)\s+{op_re}(?:\s+(.+))?$"
        m = re.match(pattern, atom_str)
        if m:
            key = m.group(1)
            raw_val = m.group(2).strip() if m.lastindex >= 2 and m.group(2) else None
            return key, op_name, _parse_rhs(raw_val)
    raise ValueError(f"Cannot parse condition atom: {atom_str!r}")


def _evaluate_atom(key: str, op: str, value: Any, variables: dict) -> bool:
    """Evaluate a single atomic condition against the variables dict."""
    left = _get_nested(variables, key)
    fn = _OP_TABLE.get(op)
    if fn is None:
        raise ValueError(f"Unknown operator: {op!r}")
    try:
        return bool(fn(left, value))
    except TypeError as exc:
        raise TypeError(
            f"Type error evaluating '{key} {op} {value!r}': {exc} "
            f"(got {type(left).__name__})"
        ) from exc


# ---------------------------------------------------------------------------
# ConditionEngine
# ---------------------------------------------------------------------------

class ConditionEngine:
    """Evaluate conditions for including or excluding dynamic prompt sections.

    Uses a simple, **safe** DSL — no ``eval()`` is invoked.

    Example::

        engine = ConditionEngine()

        engine.evaluate("plan == 'premium'", {"plan": "premium"})
        # True

        engine.evaluate(
            "plan == 'premium' AND country == 'DE'",
            {"plan": "premium", "country": "US"},
        )
        # False — country doesn't match

        print(engine.explain(
            "plan == 'premium' AND score > 0.5",
            {"plan": "premium", "score": 0.85},
        ))
    """

    # Expose for introspection / documentation
    OPERATORS = _OP_TABLE

    # ------------------------------------------------------------------
    # Core evaluation
    # ------------------------------------------------------------------

    def evaluate(self, condition: str, variables: dict) -> bool:
        """Evaluate a condition string against variables.

        Supported formats:

        - ``"query_type == 'billing'"``
        - ``"user.plan in ['premium', 'enterprise']"``
        - ``"user.country not_in ['US', 'CA']"``
        - ``"sentiment_score > 0.7"``
        - ``"user.email contains '@enterprise'"``
        - ``"conversation_history exists"``
        - ``"plan == 'premium' AND country == 'US'"``
        - ``"country == 'DE' OR country == 'FR'"``

        Args:
            condition: Condition string to evaluate.
            variables: Variable dict to evaluate against.

        Returns:
            ``True`` if the condition matches, ``False`` otherwise.
        """
        condition = condition.strip()
        # OR has lower precedence than AND.
        # Split by OR first, then by AND within each branch.
        for or_branch in _OR_SPLITTER.split(condition):
            and_atoms = _AND_SPLITTER.split(or_branch)
            try:
                branch_true = all(
                    _evaluate_atom(*_parse_atom(a.strip()), variables)
                    for a in and_atoms
                )
            except Exception as exc:
                logger.warning("Condition evaluation error in %r: %s", condition, exc)
                branch_true = False
            if branch_true:
                return True
        return False

    def evaluate_all(
        self,
        conditions: dict[str, str],
        variables: dict,
    ) -> list[str]:
        """Evaluate multiple named conditions.

        Args:
            conditions: Mapping of condition name → condition string.
            variables:  Variable dict to evaluate against.

        Returns:
            Names of conditions that evaluated to ``True``.
        """
        return [
            name
            for name, cond in conditions.items()
            if self.evaluate(cond, variables)
        ]

    # ------------------------------------------------------------------
    # Validation
    # ------------------------------------------------------------------

    def validate_condition(self, condition: str) -> tuple[bool, str]:
        """Validate a condition string for syntax errors.

        Does **not** evaluate the condition — only checks parseability.

        Args:
            condition: Condition string to validate.

        Returns:
            ``(True, "")`` if valid; ``(False, error_message)`` if not.
        """
        try:
            for or_branch in _OR_SPLITTER.split(condition):
                for atom in _AND_SPLITTER.split(or_branch):
                    _parse_atom(atom.strip())
            return True, ""
        except Exception as exc:
            return False, str(exc)

    # ------------------------------------------------------------------
    # Debugging
    # ------------------------------------------------------------------

    def explain(self, condition: str, variables: dict) -> str:
        """Explain why a condition evaluated to True or False.

        Useful for debugging "why wasn't this section included?".

        Args:
            condition: Condition string to explain.
            variables: Variable dict to evaluate against.

        Returns:
            Human-readable multi-line explanation string.
        """
        lines: list[str] = [f"Evaluating: {condition!r}"]

        or_branches = _OR_SPLITTER.split(condition)
        branch_results: list[bool] = []

        for b_idx, or_branch in enumerate(or_branches, 1):
            or_branch = or_branch.strip()
            and_atoms = _AND_SPLITTER.split(or_branch)

            if len(or_branches) > 1:
                lines.append(f"\n  [OR branch {b_idx}] {or_branch!r}")

            atom_results: list[bool] = []
            for atom_str in and_atoms:
                atom_str = atom_str.strip()
                try:
                    key, op, value = _parse_atom(atom_str)
                    left = _get_nested(variables, key)
                    result = _evaluate_atom(key, op, value, variables)
                    atom_results.append(result)
                    check = "✓" if result else "✗"
                    lines.append(
                        f"    {atom_str!r:40s}  →  "
                        f"{key}={left!r}  {op}  {value!r}  [{check}]"
                    )
                except Exception as exc:
                    atom_results.append(False)
                    lines.append(f"    {atom_str!r}  →  ERROR: {exc}")

            branch_ok = all(atom_results) if atom_results else False
            branch_results.append(branch_ok)
            if len(or_branches) > 1:
                lines.append(
                    f"    Branch result: {'True ✓' if branch_ok else 'False ✗'}"
                )

        final = any(branch_results)
        lines.append(f"\nOverall: {'True ✓' if final else 'False ✗'}")
        return "\n".join(lines)


# ---------------------------------------------------------------------------
# Demo
# ---------------------------------------------------------------------------

def _demo() -> None:
    engine = ConditionEngine()

    variables = {
        "plan": "premium",
        "country": "DE",
        "score": 0.85,
        "user": {"plan": "enterprise", "email": "alice@enterprise.com"},
    }

    print("=== Simple Conditions ===")
    simple_cases: list[tuple[str, bool]] = [
        ("plan == 'premium'",                    True),
        ("plan == 'free'",                       False),
        ("plan != 'free'",                       True),
        ("country in ['DE', 'FR', 'ES']",        True),
        ("country in ['US', 'CA']",              False),
        ("country not_in ['US', 'CA']",          True),
        ("score > 0.7",                          True),
        ("score < 0.5",                          False),
        ("plan exists",                          True),
        ("missing_key exists",                   False),
        ("user.plan == 'enterprise'",            True),
        ("user.email contains '@enterprise'",    True),
    ]
    for cond, expected in simple_cases:
        result = engine.evaluate(cond, variables)
        mark = "✓" if result == expected else "✗"
        print(f"  {mark} {cond!r:<45s} → {result}")

    print("\n=== Compound Conditions ===")
    compound_cases: list[tuple[str, bool]] = [
        ("plan == 'premium' AND country == 'DE'",  True),
        ("plan == 'premium' AND country == 'US'",  False),
        ("country == 'US' OR country == 'DE'",     True),
        ("country == 'US' OR country == 'CA'",     False),
    ]
    for cond, expected in compound_cases:
        result = engine.evaluate(cond, variables)
        mark = "✓" if result == expected else "✗"
        print(f"  {mark} {cond!r:<45s} → {result}")

    print("\n=== Explain ===")
    print(engine.explain(
        "plan == 'premium' AND score > 0.5",
        {"plan": "premium", "score": 0.85},
    ))

    print("\n=== Validation ===")
    valid_cases = [
        ("plan == 'premium'",    True),
        ("score > 0.7",          True),
        ("country in ['DE']",    True),
        ("plan is premium",      False),  # 'is' not in DSL
        ("plan",                 False),  # no operator
    ]
    for cond, expect_valid in valid_cases:
        ok, msg = engine.validate_condition(cond)
        mark = "✓" if ok == expect_valid else "✗"
        status = "valid" if ok else f"invalid: {msg}"
        print(f"  {mark} {cond!r:<30s} → {status}")

    print("\n=== evaluate_all ===")
    conditions = {
        "is_premium":       "plan == 'premium'",
        "is_eu":            "country in ['DE', 'FR', 'ES', 'IT']",
        "high_score":       "score > 0.9",     # False — score is 0.85
    }
    matched = engine.evaluate_all(conditions, variables)
    print(f"  Matched: {matched}")  # ['is_premium', 'is_eu']


if __name__ == "__main__":
    logging.basicConfig(level=logging.WARNING)
    _demo()

File: code/go/05-context-assembly/condition_engine.go
// Condition Engine — evaluate DSL conditions for dynamic prompt sections.
//
// Go port of code/python/05-context-assembly/condition_engine.py
//
// Supports a simple, safe condition DSL for including or excluding prompt
// sections based on runtime variables. No eval() is used; conditions are
// parsed with regex-based tokenisation.
//
// Simple conditions:
//
//	"plan == 'premium'"
//	"user.plan in ['premium', 'enterprise']"
//	"sentiment_score > 0.7"
//	"user.email contains '@enterprise'"
//	"conversation_history exists"
//
// Compound conditions:
//
//	"plan == 'premium' AND country == 'US'"
//	"country == 'DE' OR country == 'FR'"
//
// See: docs/04-context-engineering/02-dynamic-prompt-assembly.md
package main

import (
	"fmt"
	"log"
	"regexp"
	"strconv"
	"strings"
)

// ---------------------------------------------------------------------------
// Operator patterns — order matters: longer / multi-word first
// ---------------------------------------------------------------------------

type opPattern struct {
	re   *regexp.Regexp
	name string
}

var condOpPatterns = []opPattern{
	{regexp.MustCompile(`(?i)^([\w.]+)\s+not\s+in\s+(.+)$`), "not_in"},
	{regexp.MustCompile(`(?i)^([\w.]+)\s+not_in\s+(.+)$`), "not_in"},
	{regexp.MustCompile(`^([\w.]+)\s+>=\s+(.+)$`), "gte"},
	{regexp.MustCompile(`^([\w.]+)\s+<=\s+(.+)$`), "lte"},
	{regexp.MustCompile(`^([\w.]+)\s+==\s+(.+)$`), "eq"},
	{regexp.MustCompile(`^([\w.]+)\s+!=\s+(.+)$`), "neq"},
	{regexp.MustCompile(`^([\w.]+)\s+>\s+(.+)$`), "gt"},
	{regexp.MustCompile(`^([\w.]+)\s+<\s+(.+)$`), "lt"},
	{regexp.MustCompile(`(?i)^([\w.]+)\s+contains\s+(.+)$`), "contains"},
	{regexp.MustCompile(`(?i)^([\w.]+)\s+exists$`), "exists"},
	{regexp.MustCompile(`(?i)^([\w.]+)\s+in\s+(.+)$`), "in"},
}

var (
	andSplitter = regexp.MustCompile(`\bAND\b`)
	orSplitter  = regexp.MustCompile(`\bOR\b`)
	listItemRe  = regexp.MustCompile(`'([^']*)'|"([^"]*)"|(\S+)`)
)

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

// condGetNestedValue resolves a dotted key path from a nested map[string]any.
// "user.plan" → variables["user"]["plan"]
func condGetNestedValue(variables map[string]any, key string) any {
	parts := strings.Split(key, ".")
	var current any = variables
	for _, part := range parts {
		switch v := current.(type) {
		case map[string]any:
			current = v[part]
		default:
			return nil
		}
		if current == nil {
			return nil
		}
	}
	return current
}

// parseRHS parses the right-hand side of a condition into a typed value.
// Handles: 'string', "string", numbers, booleans, and ['a','b','c'] lists.
func parseRHS(raw string) any {
	raw = strings.TrimSpace(raw)
	if raw == "" {
		return nil
	}

	// Single-quoted string
	if strings.HasPrefix(raw, "'") && strings.HasSuffix(raw, "'") {
		return raw[1 : len(raw)-1]
	}
	// Double-quoted string
	if strings.HasPrefix(raw, "\"") && strings.HasSuffix(raw, "\"") {
		return raw[1 : len(raw)-1]
	}
	// List literal: ['a', 'b'] or ["a", "b"]
	if strings.HasPrefix(raw, "[") && strings.HasSuffix(raw, "]") {
		inner := raw[1 : len(raw)-1]
		var list []any
		for _, m := range listItemRe.FindAllStringSubmatch(inner, -1) {
			if m[1] != "" {
				list = append(list, m[1])
			} else if m[2] != "" {
				list = append(list, m[2])
			} else if m[3] != "" {
				list = append(list, m[3])
			}
		}
		return list
	}
	// Boolean
	if raw == "true" {
		return true
	}
	if raw == "false" {
		return false
	}
	// Integer
	if i, err := strconv.ParseInt(raw, 10, 64); err == nil {
		return float64(i)
	}
	// Float
	if f, err := strconv.ParseFloat(raw, 64); err == nil {
		return f
	}
	return raw
}

// parseAtom parses one atomic condition string into (key, opName, value).
func parseAtom(atomStr string) (key, opName string, value any, err error) {
	atomStr = strings.TrimSpace(atomStr)
	for _, p := range condOpPatterns {
		m := p.re.FindStringSubmatch(atomStr)
		if m == nil {
			continue
		}
		key = m[1]
		opName = p.name
		if len(m) > 2 {
			value = parseRHS(m[2])
		}
		return key, opName, value, nil
	}
	return "", "", nil, fmt.Errorf("cannot parse condition atom: %q", atomStr)
}

// condEvaluateAtom evaluates one atomic condition against variables.
func condEvaluateAtom(key, op string, value any, variables map[string]any) (bool, error) {
	left := condGetNestedValue(variables, key)

	switch op {
	case "exists":
		return left != nil, nil
	case "eq":
		return condCompareValues(left, value) == 0, nil
	case "neq":
		return condCompareValues(left, value) != 0, nil
	case "gt":
		cmp, err := condNumericCompare(left, value)
		return err == nil && cmp > 0, err
	case "lt":
		cmp, err := condNumericCompare(left, value)
		return err == nil && cmp < 0, err
	case "gte":
		cmp, err := condNumericCompare(left, value)
		return err == nil && cmp >= 0, err
	case "lte":
		cmp, err := condNumericCompare(left, value)
		return err == nil && cmp <= 0, err
	case "in":
		return condValueInList(left, value), nil
	case "not_in":
		return !condValueInList(left, value), nil
	case "contains":
		ls, ok1 := condToString(left)
		rs, ok2 := condToString(value)
		if !ok1 || !ok2 {
			return false, nil
		}
		return strings.Contains(ls, rs), nil
	}
	return false, fmt.Errorf("unknown operator: %q", op)
}

func condToString(v any) (string, bool) {
	switch s := v.(type) {
	case string:
		return s, true
	}
	return "", false
}

func condToFloat(v any) (float64, bool) {
	switch n := v.(type) {
	case float64:
		return n, true
	case int:
		return float64(n), true
	case int64:
		return float64(n), true
	case string:
		if f, err := strconv.ParseFloat(n, 64); err == nil {
			return f, true
		}
	}
	return 0, false
}

func condNumericCompare(a, b any) (int, error) {
	fa, okA := condToFloat(a)
	fb, okB := condToFloat(b)
	if !okA || !okB {
		return 0, fmt.Errorf("cannot compare %T and %T numerically", a, b)
	}
	switch {
	case fa < fb:
		return -1, nil
	case fa > fb:
		return 1, nil
	default:
		return 0, nil
	}
}

// condCompareValues returns 0 if equal, non-zero otherwise (string or numeric).
func condCompareValues(a, b any) int {
	if cmp, err := condNumericCompare(a, b); err == nil {
		return cmp
	}
	sa, _ := condToString(a)
	sb, _ := condToString(b)
	if sa == sb {
		return 0
	}
	return -1
}

func condValueInList(value, list any) bool {
	items, ok := list.([]any)
	if !ok {
		return false
	}
	for _, item := range items {
		if condCompareValues(value, item) == 0 {
			return true
		}
	}
	return false
}

// ---------------------------------------------------------------------------
// ConditionEngine
// ---------------------------------------------------------------------------

// ConditionEngine evaluates conditions for dynamic prompt sections.
// Uses a simple, safe DSL — no eval() is invoked.
type ConditionEngine struct{}

// NewConditionEngine creates a new ConditionEngine.
func NewConditionEngine() *ConditionEngine {
	return &ConditionEngine{}
}

// Evaluate evaluates a condition string against variables.
//
// Supported formats:
//   - "query_type == 'billing'"
//   - "user.plan in ['premium', 'enterprise']"
//   - "user.country not_in ['US', 'CA']"
//   - "sentiment_score > 0.7"
//   - "user.email contains '@enterprise'"
//   - "conversation_history exists"
//   - "plan == 'premium' AND country == 'US'"
//   - "country == 'DE' OR country == 'FR'"
func (ce *ConditionEngine) Evaluate(condition string, variables map[string]any) bool {
	condition = strings.TrimSpace(condition)
	// OR has lower precedence than AND.
	for _, orBranch := range orSplitter.Split(condition, -1) {
		andAtoms := andSplitter.Split(orBranch, -1)
		branchTrue := true
		for _, atom := range andAtoms {
			atom = strings.TrimSpace(atom)
			key, op, value, err := parseAtom(atom)
			if err != nil {
				log.Printf("condition_engine: parse error in %q: %v", condition, err)
				branchTrue = false
				break
			}
			result, err := condEvaluateAtom(key, op, value, variables)
			if err != nil {
				log.Printf("condition_engine: eval error in %q: %v", condition, err)
				branchTrue = false
				break
			}
			if !result {
				branchTrue = false
				break
			}
		}
		if branchTrue {
			return true
		}
	}
	return false
}

// EvaluateAll evaluates multiple named conditions and returns those that are true.
func (ce *ConditionEngine) EvaluateAll(conditions map[string]string, variables map[string]any) []string {
	var matched []string
	for name, cond := range conditions {
		if ce.Evaluate(cond, variables) {
			matched = append(matched, name)
		}
	}
	return matched
}

// ValidateCondition validates a condition string for syntax errors.
// Returns (true, "") if valid; (false, errorMessage) if not.
func (ce *ConditionEngine) ValidateCondition(condition string) (bool, string) {
	for _, orBranch := range orSplitter.Split(condition, -1) {
		for _, atom := range andSplitter.Split(orBranch, -1) {
			_, _, _, err := parseAtom(strings.TrimSpace(atom))
			if err != nil {
				return false, err.Error()
			}
		}
	}
	return true, ""
}

// Explain returns a human-readable explanation of why a condition evaluated to true or false.
func (ce *ConditionEngine) Explain(condition string, variables map[string]any) string {
	var lines []string
	lines = append(lines, fmt.Sprintf("Evaluating: %q", condition))

	orBranches := orSplitter.Split(condition, -1)
	var branchResults []bool

	for bIdx, orBranch := range orBranches {
		orBranch = strings.TrimSpace(orBranch)
		andAtoms := andSplitter.Split(orBranch, -1)

		if len(orBranches) > 1 {
			lines = append(lines, fmt.Sprintf("\n  [OR branch %d] %q", bIdx+1, orBranch))
		}

		var atomResults []bool
		for _, atomStr := range andAtoms {
			atomStr = strings.TrimSpace(atomStr)
			key, op, value, err := parseAtom(atomStr)
			if err != nil {
				atomResults = append(atomResults, false)
				lines = append(lines, fmt.Sprintf("    %q  →  ERROR: %v", atomStr, err))
				continue
			}
			left := condGetNestedValue(variables, key)
			result, err := condEvaluateAtom(key, op, value, variables)
			if err != nil {
				atomResults = append(atomResults, false)
				lines = append(lines, fmt.Sprintf("    %q  →  ERROR: %v", atomStr, err))
				continue
			}
			atomResults = append(atomResults, result)
			check := "✓"
			if !result {
				check = "✗"
			}
			lines = append(lines, fmt.Sprintf("    %-40q  →  %s=%v  %s  %v  [%s]",
				atomStr, key, left, op, value, check))
		}

		branchOk := len(atomResults) > 0
		for _, r := range atomResults {
			if !r {
				branchOk = false
				break
			}
		}
		branchResults = append(branchResults, branchOk)
		if len(orBranches) > 1 {
			status := "True ✓"
			if !branchOk {
				status = "False ✗"
			}
			lines = append(lines, fmt.Sprintf("    Branch result: %s", status))
		}
	}

	final := false
	for _, r := range branchResults {
		if r {
			final = true
			break
		}
	}
	overall := "True ✓"
	if !final {
		overall = "False ✗"
	}
	lines = append(lines, fmt.Sprintf("\nOverall: %s", overall))
	return strings.Join(lines, "\n")
}
